docs/subnets/create-a-vm-blobvm.md

```diff
@@ -0,0 +1,454 @@
+
+```go
+// Returning a new VM instance from VM's factory
+func (f *Factory) New(*snow.Context) (interface{}, error) { return &vm.VM{}, nil }
+```
+
+VM's methods and other parts of the code are made available to `avalanchego` by **registering** its factory and instantiating a new VM using this factory whenever a chain is required to be built. It's generally in the [`factory.go`](https://github.com/ava-labs/blobvm/blob/master/factory.go) file of the VM.
+
+```go
+// Instantiating a new VM from its factory in avalanchego
+func (m *manager) buildChain(chainParams ChainParameters, sb Subnet) (*chain, error) {
+    ...
+    vmFactory, err := m.VMManager.GetFactory(vmID)
+    vm, err := vmFactory.New(ctx.Context)
+    ...
+}
+```
+
+### Initializing a VM
+
+Chains are functional, only when the instantiated VMs are initialized. Initializing a VM involves setting up the database, block builder, mempool, genesis state, handlers, etc. This will expose the VM's API handlers, start accepting transactions, gossiping them across the network, building blocks, etc.
+
+```go
+if err := vm.Initialize(
+    ctx.Context,
+    vmDBManager,
+    genesisData,
+    chainConfig.Upgrade,
+    chainConfig.Config,
+    msgChan,
+    fxs,
+    sender,
+);
+```
+
+You can refer to the implementation of `vm.initialize` in the [BlobVM](https://github.com/ava-labs/blobvm/blob/1fbb655246a2a41861f40d603ad40299b177252a/vm/vm.go#L92) repository.
+
+### Interaction
+
+Once a VM is initialized, it becomes a functional chain. We can use the handlers to issue transactions, query chain state, and whatever functionalities are being provided by the VM.
+
+### Consensus Engine
+
+## Components of a Virtual Machine
+
+We have to implement the following components in our virtual machine. 
+
+* **Transaction** - Transaction structure, initialization, execution etc.
+* **Mempool** - Heap (min and max) for keeping pending transactions locally on the node.
+* **Network** - Implements gossip of transactions from mempool to the network.
+* **Block** - It handles block creation, initialization, verification, parsing, etc.
+* **Builder** - Returns the built preferred block by including transactions from the mempool.
+* **Block Builder** - Initiates the gossip of new transactions, notifies the engine that new blocks are ready to be built, and also handles next block generation notification.
+* **Storage** - To store and retrieve data of the chain's state.
+* **Virtual Machine** - Entry point for all the components to orchestrate them according to the consensus engine and API requests.
+* **Service** - API handlers for interacting with VM and initialized chain.
+* **Client** - It is used for interacting with service APIs.
+* **Factory** - For creating new instances of the virtual machine.
+
+## Transaction Lifecycle
```

Comment:
I'm a bit skeptical on if we need to go so deep in this tutorial on how avalanche handles your transactions under-the-hood. I feel like user's should just be concerned with how to implement a VM + transaction types and how avalanche handles it should just be a black-box that they don't care about.

Reply:
Isn't transaction lifecycle defined by a VM? I have used VM components to explain how the transaction is processed within.

Reply:
Transactions are internal to the VM, the consensus engine only cares about blocks, so it's not necessary to go through it. The phrasing in this document suggests transactions are a hard requirement, so we should make sure that the flexibility of the VM interface is clear (or at least that we don't say that it's more rigid than it is).

Reply:
Got it. Made it clear in the docs.